pandas/tests/io/excel/test_writers.py

```diff
@@ -1115,6 +1115,27 @@ def test_bytes_io(self, engine):
             reread_df = pd.read_excel(bio, index_col=0)
             tm.assert_frame_equal(df, reread_df)
 
+    def test_engine_kwargs(self, engine, path):
+        # GH#52368
+        df = DataFrame([{"A": 1, "B": 2}, {"A": 3, "B": 4}])
+
+        msgs = {
+            "odf": r"OpenDocumentSpreadsheet() got an unexpected keyword "
+            r"argument 'foo'",
+            "openpyxl": r"load_workbook() got an unexpected keyword argument 'foo'",
+            "xlsxwriter": r"__init__() got an unexpected keyword argument 'foo'",
+        }
+
+        # Handle change in error message for openpyxl (write and append mode)
+        if engine == "openpyxl" and os.path.exists(path):
+            msgs["openpyxl"] = r"__init__() got an unexpected keyword argument 'foo'"
```

Comment:
For xlsxwriter and the append mode of openpyxl, I would not expect this error message. It appears that `foo` is attempting to be passed to `OpenpyxlWriter.__init__` rather than stored in engine_kwargs. Is that the case?

Reply:
@rhshadrach For `openpyxl.load_workbook`, we are getting the correct error message of `load_workbook() got an unexpected keyword argument 'foo'`


For xlsxwriter (there is no append mode) and the write mode of openpyxl, both methods are passed to <engine>.Workbook(). 

![238488486-9a5a8058-aae1-4f2f-936d-fc04d01346f9](https://github.com/pandas-dev/pandas/assets/45905457/3f8943c6-b9dc-45db-8bae-a753a59869eb)

Reply:
@rhshadrach Yeah looking at the workbook class being instantiated in both `_xlsxwriter.py` and `_openpyxl.py`, the error message handling should be correct.

Reply:
I see - the `__init__` that raises is part of the engine. Can you check the message for e.g. `Workbook.__init__()` to make sure we're not raising in the wrong place.

Reply:
Sure!

Reply:
Looks like the traceback was changed in Python 3.10; in Python 3.9 you only get `__init__()...`. Can you import `PY310` from pandas.compat._constants and check for `Workbook.__init__` when true? Note this will be true whenever the Python version is 3.10 or greater.

Reply:
Thanks @rhshadrach! I was going down a rabbit hole last night troubleshooting this. I was beginning to think that the issue had something to do with how we were raising the Type error in `_openpyxl.py` and `_xlsxwriter.py` without an [Exception clause](https://stackoverflow.com/questions/18001721/raise-with-no-argument).